results/pkg/watcher/reconciler/pipelinerun/pipelinerun_test.go

```diff
@@ -0,0 +1,137 @@
+	"knative.dev/pkg/configmap"
+)
+
+func TestReconcile(t *testing.T) {
+	pipelineRunTest := NewPipelineRunTest(t)
+
+	testFuncs := map[string]func(t *testing.T){
+		"Create":   pipelineRunTest.testCreatePipelineRun,
+		"Unchange": pipelineRunTest.testUnchangePipelineRun,
+		"Update":   pipelineRunTest.testUpdatePipelineRun,
+	}
+
+	for name, testFunc := range testFuncs {
+		t.Run(name, testFunc)
+	}
+}
+
+type PipelineRunTest struct {
+	pipelineRun *v1beta1.PipelineRun
+	asset       test.Assets
+	ctx         context.Context
+	client      pb.ResultsClient
+}
+
+func NewPipelineRunTest(t *testing.T) PipelineRunTest {
+	client := common.NewResultsClient(t)
+	pipelineRun := &v1beta1.PipelineRun{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:        "Tekton-PipelineRun",
+			Namespace:   "default",
+			Annotations: map[string]string{"demo": "pipelinerun_demo"},
+			UID:         "54321",
+		},
+	}
+	asset, ctx := getFakeClients(t, []*v1beta1.PipelineRun{pipelineRun}, client)
+	pipelineRunTest := PipelineRunTest{
+		pipelineRun: pipelineRun,
+		asset:       asset,
+		ctx:         ctx,
+		client:      client,
+	}
+	return pipelineRunTest
+}
+
+func (tt *PipelineRunTest) testCreatePipelineRun(t *testing.T) {
+	pr, err := common.ReconcilePipelineRun(tt.ctx, tt.asset, tt.pipelineRun)
+	if err != nil {
+		t.Fatalf("Failed to get completed PipelineRun %s: %v", tt.pipelineRun.Name, err)
+	}
+	if _, ok := pr.Annotations[common.IDName]; !ok {
+		t.Fatalf("Expected completed PipelineRun %s should be updated with a results_id field in annotations", tt.pipelineRun.Name)
+	}
+	if _, err := tt.client.GetResult(tt.ctx, &pb.GetResultRequest{Name: pr.Annotations[common.IDName]}); err != nil {
+		t.Fatalf("Expected completed PipelineRun %s not created in api server", tt.pipelineRun.Name)
+	}
+}
+
+func (tt *PipelineRunTest) testUnchangePipelineRun(t *testing.T) {
+	pr, err := tt.asset.Clients.Pipeline.TektonV1beta1().PipelineRuns(tt.pipelineRun.Namespace).Get(tt.pipelineRun.Name, metav1.GetOptions{})
+	if err != nil {
+		t.Fatalf("Failed to get completed PipelineRun %s: %v", tt.pipelineRun.Name, err)
+	}
+	newpr, err := common.ReconcilePipelineRun(tt.ctx, tt.asset, pr)
+	if err != nil {
+		t.Fatalf("Failed to get completed PipelineRun %s: %v", tt.pipelineRun.Name, err)
+	}
+	if diff := cmp.Diff(pr, newpr); diff != "" {
+		t.Fatalf("Expected completed PipelineRun should remain unchanged when it has a results_id in annotations: %v", diff)
+	}
+}
+
+func (tt *PipelineRunTest) testUpdatePipelineRun(t *testing.T) {
+	pr, err := common.ReconcilePipelineRun(tt.ctx, tt.asset, tt.pipelineRun)
+	if err != nil {
+		t.Fatalf("Failed to get completed PipelineRun %s: %v", tt.pipelineRun.Name, err)
+	}
+	pr.UID = "234435"
+	if _, err = tt.asset.Clients.Pipeline.TektonV1beta1().PipelineRuns(tt.pipelineRun.Namespace).Update(pr); err != nil {
+		t.Fatalf("Failed to update PipelineRun %s: %v", tt.pipelineRun.Name, err)
+	}
+	updatepr, err := common.ReconcilePipelineRun(tt.ctx, tt.asset, pr)
```

Comment:
Is there a purpose to the double Reconcile here, or is 1 sufficient?

Reply:
The one in the first line is to get the newest version of pr from both the result client and the pipeline client, and the second one is for testing the `update` operation.